### DISCUSSION TITLE
How to get the actual dependency chain in manualChunks

### DISCUSSION THREAD

**Discussion by susanforme** [rollup/rollup#5745] [Q&A]
Purpose: I need to split modules based on the entry points. Static imports should be bundled into a vendor chunk, and dynamic imports should be handled differently.

Description: In manualChunks, when using meta.getModuleInfo(id) to get the importedIds, it includes the utils.ts module. However, when I use meta.getModuleInfo("utils.ts"), the isIncluded flag is false (indicating it was tree-shaken). In reality, the main.ts file statically imports rename.ts, and expensiveWork is statically imported by asyncModule and then dynamically imported by main.ts. How can I obtain this actual import chain?

Given Files

main.ts

```ts
import {rename} from './utils'

async function foo(){
    await import("./asyncModule")
}
```

asyncModule

```ts
import {expensiveWork} from './utils';
expensiveWork()
```

utils.ts

```ts
export * from './rename'
export * from './expensiveWork'
```

rename.ts

```ts
export function rename(){
    
}
```

expensiveWork.ts

```ts
export function expensiveWork(){
    
}
```

**Comment by chrisjcthomas** [comment]
`manualChunks` runs while Rollup is still figuring out the final module graph. A few of the flags/APIs are easy to misread in the situation you described.

### Why `importedIds` shows `utils` but `isIncluded` is false

- `importedIds` / `dynamicallyImportedIds` are **static resolution** results: “this module has an import record pointing at X”.
- `isIncluded` means “this module contributes something to the **final** output after tree-shaking”.

Barrel files make that diverge a lot:

```ts
// utils.ts
export * from './rename'
export * from './expensiveWork'
```

`main.ts` only needs `rename`. After linking, Rollup can re-export / side-step the barrel so `utils.ts` itself is not emitted (`isIncluded === false`) while `rename.ts` is included. `asyncModule` pulls `expensiveWork` through the same barrel; that edge is on a **dynamic** import path from `main`, so it belongs in a different chunk decision than the static `rename` path.

So: `getModuleInfo('utils.ts').isIncluded === false` can be correct even when something “imports utils” in source.

### How to walk the real used graph

Prefer **included** modules and follow **both** static and dynamic edges:

```js
function manualChunks(id, { getModuleInfo, getModuleIds }) {
  const info = getModuleInfo(id)
  if (!info || !info.isIncluded) return

  // importers = modules that statically import this one
  // dynamicImporters = modules that import() it
  const staticEntry = info.importers.some((imp) => getModuleInfo(imp)?.isEntry)
  const dynamicOnly =
    info.dynamicImporters.length > 0 && info.importers.length === 0

  if (id.includes("node_modules")) {
    return dynamicOnly ? "async-vendor" : "vendor"
  }

  // example: put modules only reached via import() into async-*
  if (dynamicOnly) return "async-app"
}
```

Useful fields on `ModuleInfo`:

| Field | Meaning |
| --- | --- |
| `importers` | static parents |
| `dynamicImporters` | `import()` parents |
| `importedIds` | static children (resolved ids) |
| `dynamicallyImportedIds` | dynamic children |
| `isIncluded` | survives tree-shaking |
| `isEntry` | entry module |

To answer “is this module on a static path from entry E?” walk `importers` upward only through `isIncluded` nodes. To answer “is it only needed for a dynamic import?” check that every path from an entry hits a `dynamicallyImportedIds` edge before reaching it (or simply `importers.length === 0 && dynamicImporters.length > 0` for the leaf case).

### Barrel-specific tip

For chunking by “who uses which export”, barrels fight you. Prefer direct imports in app code:

```ts
import { rename } from './rename'
// dynamic path
const { expensiveWork } = await import('./expensiveWork')
```

Then `manualChunks` sees separate modules with clear static vs dynamic importers, and you no longer care that `utils.ts` was optimized out.

### Practical recipe for your goal

> Static imports → vendor/shared chunk; dynamic imports handled differently

```js
manualChunks(id, { getModuleInfo }) {
  if (!id.includes("node_modules")) return

  const { importers, dynamicImporters, isIncluded } = getModuleInfo(id) || {}
  if (!isIncluded) return

  const onlyDynamic = dynamicImporters?.length && !importers?.length
  return onlyDynamic ? "vendor-async" : "vendor"
}
```

If a module is imported both statically and dynamically, Rollup must put it in a place both parents can load — usually the static/shared chunk. That is expected; you cannot keep a single module “dynamic-only” once a static entry also depends on it.

Bottom line: trust `isIncluded` + `importers` / `dynamicImporters` over raw `importedIds` on barrels, and split chunk names from those edges rather than from the intermediate re-export file.
